gradient_accumulator/agc.py

```diff
@@ -1,81 +1,88 @@
+SCALAR = tf.constant([], dtype=tf.int32)
+LINEAR = tf.constant([0], dtype=tf.int32)
+TENSOR2D = tf.constant([0, 1], dtype=tf.int32)
+TENSOR3D = tf.constant([0, 1, 2], dtype=tf.int32)
+TENSOR4D = tf.constant([0, 1, 2, 3], dtype=tf.int32)
+
+
+@tf.function
 def compute_norm(x, axis, keepdims):
     """
     Computes the euclidean norm of a tensor :math:`x`.
-
-    Args:
-        x: input tensor.
-        axis: which axis to compute norm across.
-        keepdims: whether to keep dimension after applying along axis.
-
-    Returns:
-        Euclidean norm.
     """
-    return tf.math.reduce_sum(x**2, axis=axis, keepdims=keepdims) ** 0.5
+    return tf.sqrt(tf.reduce_sum(tf.square(x), axis=axis, keepdims=keepdims))
 
 
+@tf.function
 def unitwise_norm(x):
     """
-    Wrapper class which dynamically sets `axis` and `keepdims` given an
-    input `x` for calculating euclidean norm.
+    Computes the unitwise norm of a tensor.
+    """
 
-    Args:
-        x: input tensor.
+    def compute_reduction_axes(r):
+        axes = tf.case(
+            [
+                (
+                    tf.equal(r, 1),
+                    lambda: SCALAR,
+                ),
+                (
+                    tf.equal(r, 2),
+                    lambda: LINEAR,
+                ),
+                (
+                    tf.equal(r, 3),
+                    lambda: TENSOR2D,
+                ),
+                (
+                    tf.equal(r, 4),
+                    lambda: TENSOR3D,
+                ),
+                (
+                    tf.equal(r, 5),
+                    lambda: TENSOR4D,
+                ),
+            ],
+            default=lambda: SCALAR,
+        )
+        return axes
 
-    Returns:
-        Euclidean norm.
-    """
-    if len(x.get_shape()) <= 1:  # Scalars and vectors
-        axis = None
-        keepdims = False
-    elif len(x.get_shape()) in [
-        2,
-        3,
-    ]:  # Linear layers of shape IO or multihead linear
-        axis = 0
-        keepdims = True
-    elif len(x.get_shape()) == 4:  # Conv kernels of shape HWIO
-        axis = [0, 1, 2]
-        keepdims = True
-    elif len(x.get_shape()) == 5:  # Conv kernels of shape HWDIO
-        axis = [0, 1, 2, 3]
-        keepdims = True
-    else:
-        raise ValueError(f"Got a parameter with shape not in [1, 2, 4, 5]! {x}")
-    return compute_norm(x, axis, keepdims)
+    return compute_norm(
+        x, axis=compute_reduction_axes(tf.rank(x)), keepdims=True
+    )
 
 
+@tf.function
 def adaptive_clip_grad(
     parameters, gradients, clip_factor: float = 0.01, eps: float = 1e-3
 ):
     """
     Performs adaptive gradient clipping on a given set of parameters and
     gradients.
+    """
 
-    * Official JAX implementation (paper authors):
-      https://github.com/deepmind/deepmind-research/tree/master/nfnets  # noqa
-    * Ross Wightman's implementation
-      https://github.com/rwightman/pytorch-image-models/blob/master/timm/utils/agc.py  # noqa
```

Comment:
I'd like to hear the argument as to why you are removing documentations for these methods.

Reply:
See above